### ISSUE TITLE
Allow Bundler's deployment mode setting to be overridden

### ISSUE THREAD

**Issue by nicholaspufal** [ruby/setup-ruby#292]
Hey there, I would like to re-surface [this discussion](https://github.com/ruby/setup-ruby/issues/153) which has been closed.

Quick recap that at the moment when using:
```
with:
  bundler-cache: true # runs 'bundle install' and caches installed gems automatically
```
Bundler will run in deployment mode "true"

I recommend re-considering a mechanism to toggle that and be able to set it to "false". 

This inability to set it to "false" is currently causing problems in a GitHub Actions workflow we have in our org. 

We have dependabot setup and in any PRs created through that automation the job that's using `ruby/setup-ruby` breaks with exit code 16 coming from Bundler. That particular job doesn't care if there is a mismatch between Gemfile and Gemfile.lock — it serves another purpose and only uses Bundler to install a needed dependency. 

Like others suggested in that other thread, currently there doesn't seem to be a way to override this behavior. I've tried to set the `BUNDLE_DEPLOYMENT` env var but `bundle config --local deployment true` defined from `ruby/setup-ruby` takes precedence.

**Comment by eregon** [comment]
Deployment mode is only if there is a checked-in Gemfile.lock BTW.

Could you post a link of such a dependent PR failing?

**Comment by nicholaspufal** [comment]
> Deployment mode is only if there is a checked-in Gemfile.lock BTW.
We commit the lock file in our repos to have a more deterministic environment

> BTW it should just work with dependabot if it correctly updates the Gemfile.lock.
That's the exact underlying issue here and why I think it makes sense to allow that flag to be overridden.

We have seen Dependabot making unexpected changes to Gemfile.lock [1] and that behavior + this deployment mode always set to true they create a situation where one CI job can't run `bundle install` to install a dependency (the problem compounds because that's a job required in order to merge). 

I could unblock us by not leaning on `bundler-cache: true` but like the README in this repo suggests, caching gems can be a tricky thing so would rather not go that route if possible.

While I understand the root cause here is Dependabot there should be an easy way to unblock users of this GH action even though "deployment mode" is a sensible default. Users who opt to do that would be doing so consciously through an env variable that sets it to false for instance.


[1] To expand a bit on this issue Dependabot is causing these breaking changes to `Gemfile.lock`. The working lock file looks like this:
```
GEM
  remote: https://some.repository/gems/
  remote: https://another.repository/gems/
  specs:
    # list of gems
```

When Dependabot changes kick in it becomes like this in the PR: 
```
GEM
  remote: https://another.repository/gems/
  specs:
    # list of gems

GEM
  remote: https://some.repository/gems/
  specs:
    # cherry-picked a gem from the list of gems above
```

Which results in bundler returning that exit code only when in deployment mode

**Comment by nicholaspufal** [comment]
Sorry for taking long to reply back. Thank you for sharing your thoughts — after digging further we verified that was indeed a mismatch between Bundler versions. Dependabot was using a newer Bundler than ours (2.1.4), so your message was spot on. 

I think it would be "nice to have" a way to override that deployment mode for specific cases like this where it could be used as a temporary bandaid to unlock workflows until the proper fix can be done. 

For example: in our case upgrading Bundler requires some coordination between teams and it's not something that likely will be addressed immediately unfortunately, so in the meanwhile would be nice to unblock those PRs by overriding that deployment mode flag.

**Comment by eregon** [comment]
Disabling this would mean a checked-in Gemfile.lock would effectively be ignored.
That would be way too confusing, so I think that's not an option at all.
So far it seems all issues related to this are actually fundamental which need solving, and disabling would just fail later in more weird ways.

**Comment by eregon** [comment]
A workaround is to `rm Gemfile.lock` if one wants to ignore the checked-in Gemfile.lock.
